### ISSUE TITLE
[RFC] Remove use of MessageLogger.h from *DataFormats headers

### ISSUE THREAD

**Issue by Dr15Jones** [cms-sw/cmssw#32844]
The use of the MessageLogger is an implementation detail and therefore does not need to be exposed in the header of a *DataFormats class. Not allowing MessageLogger.h in *DataFormats headers has several advantages

- MessageLogger.h is fairly large and therefore contributes to compilation time,
- Code generated from MessageLogger calls is quite large and not suitable to inlined functions,
- Good large scale software design requires avoiding unnecessary code dependencies,
- For the moment, ROOT's module build has problems parsing the MessageLogger.h headers so removing them would allow finding potentially other problems.

**Comment by makortel** [comment]
assign core

**Comment by makortel** [comment]
@davidlange6 @vgvassilev Would you have any comments from the ROOT module point of view?

**Comment by Dr15Jones** [comment]
@davidlange6 FYI

**Comment by vgvassilev** [comment]
@makortel, I think that would make our modules infrastructure a bit more stable. Thanks @Dr15Jones for doing that!!

**Comment by vgvassilev** [comment]
Thanks, seem reasonable. We have changed the include graph in header files by pushing the inclusion of MessageLogger.h towards the edges on the edges. I see that #32847 chose to throw instead of logging an error -- can we consistently apply this rule to #32824 or there are packages that are not compiled by exceptions on? [Alternatively, we could move that one function in a different place and use it from there].

**Comment by Dr15Jones** [comment]
@vgvassilev For places which actually use the MessageLogger, the idea would be to move the calls to the .cc file instead of in the header. Changing to throwing an exception would be a much more disruptive change as it could cause major problems in production if something suddenly starts to throw.

**Comment by vgvassilev** [comment]
Yeah, that’s a better approach indeed if the functions are not relying on being inlined in the includers. I assume if they log (even conditionally) performance/inlining is not a top priority.

**Comment by Dr15Jones** [comment]
See https://github.com/cms-sw/cmssw/pull/32859
